// tailwindcss-3d/css-utilities/transform-box.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "TransformBox", {
    enumerable: true,
    get: function() {
        return TransformBox;
    }
});
var _assert_this_initialized = require("@swc/helpers/_/_assert_this_initialized");
var _class_call_check = require("@swc/helpers/_/_class_call_check");
var _define_property = require("@swc/helpers/_/_define_property");
var _inherits = require("@swc/helpers/_/_inherits");
var _create_super = require("@swc/helpers/_/_create_super");
var _base = require("./base");
var TransformBox = /*#__PURE__*/ function(Base) {
    "use strict";
    _inherits._(TransformBox, Base);
    var _super = _create_super._(TransformBox);
    function TransformBox() {
        _class_call_check._(this, TransformBox);
        var _this;
        _this = _super.apply(this, arguments);
        _define_property._(_assert_this_initialized._(_this), "utilities", function() {
            _this.api.matchUtilities({
                "transform-box": function(value) {
                    return {
                        "transform-box": value
                    };
                }
            }, {
                values: _this.api.theme("transformBox")
            });
        });
        return _this;
    }
    return TransformBox;
}(_base.Base);
_define_property._(TransformBox, "defaultTheme", {
    content: "content-box",
    border: "border-box",
    fill: "fill-box",
    stroke: "stroke-box",
    view: "view-box"
});
